### PR TITLE
Fix jerrycans disappearing when filling empty buckets, closes #4482

### LINKED ISSUE #4482: [1.16.4] Jerrycans disappear when filling buckets
### Description of the issue:
When filling empty buckets from a jerrycan, the jerrycan disappears after one bucket gets filled. This happens regardless of the can's fill state.

### Versions & Modlist
IE 1.16.4-4.1.1-128
Forge 35.1.6 and 35.1.0, latest and recommended respectively at time of writing

### PR DESCRIPTION
**Pull request by ConductiveFoam**
I *hope* this is an acceptable fix; it seems to me that the person writing it before just got the operator inverted.

### CODE REVIEW

**Review by malte0811 — CHANGES_REQUESTED** [review]
That fix is definitely not correct. `getFluidHandler` returns a `LazyOptional`, which (both by convention and in practice) is non-null. So comparing that value to null is meaningless. The proper fix would probably be something involving `LazyOptional.isPresent()`.

Make sure to also test your fix with filling a jerrycan from another jerrycan, that case currently seems to be broken as well.

**Comment by ConductiveFoam** [comment]
Came up with a different implementation for `getRemainingItems`: Keep only the first recognized Jerrycan, since that's the one that `getCraftingResult` uses as a source of fluid. All other items get discarded as before; this is also is why 9515e4d restricts recipes to two items, before one could lose items by filling more of the slots.
Now Jerrycan-Jerrycan filling works too

Hope this is better!

File: src/main/java/blusunrize/immersiveengineering/common/crafting/JerrycanRefillRecipe.java [hunk 1]
```diff
@@ -87,6 +87,16 @@ else if(ret[1].isEmpty()&&FluidUtil.getFluidHandler(stackInSlot).isPresent())
 		return ret;
 	}
 
+	private int countOccupiedSlots(IInventory inv) {
+		int c = 0;
+		for(int i = 0; i < inv.getSizeInventory(); i++)
+		{
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/crafting/JerrycanRefillRecipe.java]: Formatting (the braces should be removed)

File: src/main/java/blusunrize/immersiveengineering/common/crafting/JerrycanRefillRecipe.java [hunk 2]
```diff
@@ -87,6 +87,16 @@ else if(ret[1].isEmpty()&&FluidUtil.getFluidHandler(stackInSlot).isPresent())
 		return ret;
 	}
 
+	private int countOccupiedSlots(IInventory inv) {
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/crafting/JerrycanRefillRecipe.java]: Formatting (brace should be on the next line)

File: src/main/java/blusunrize/immersiveengineering/common/crafting/JerrycanRefillRecipe.java [hunk 3]
```diff
@@ -98,12 +108,18 @@ public boolean canFit(int width, int height)
...
-			if(FluidUtil.getFluidHandler(inv.getStackInSlot(i))!=null)
+			ItemStack stackInSlot = inv.getStackInSlot(i);
+			if(!stackInSlot.isEmpty())
 			{
+				if(Misc.jerrycan.equals(stackInSlot.getItem())&&!foundJerrycan)
+				{
+					foundJerrycan = true;
+					continue;
+				}
 				remaining.set(i, ItemStack.EMPTY);
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/crafting/JerrycanRefillRecipe.java]: IMO something like
```java
if (…)
    foundJerrycan = true;
else
    remaining.set(i, ItemStack.EMPTY);
```
would be nicer here.

**Review by malte0811 — COMMENTED** [review]
Looks good now (aside from some minor formatting problems), and seems to fix the problem it's intended to fix. There's another problem with this recipe (filling by non-1000-mB-amounts is handled by essentially using a side-channel, which seems to be broken), but that's not related to the problem described in #4482. I'll look into that problem later, unless you want to include that in this PR?

**Comment by ConductiveFoam** [comment]
Glad it looks good!
Fixed the formatting problems.

I don't get what other problem you are referring to, filling odd amounts works for me. The tooltip in the crafting grid sometimes is off, the items themselves get processed fine however. Can you elaborate? I'm happy to help out some more

**Comment by BluSunrize** [comment]
...shit I merged this into the wrong branch

**Pull request merged**